functorch/experimental/_scan.py

```diff
@@ -0,0 +1,370 @@
+            _, grads = joint(list(init) + list(xs),
+                             [grad for grad in grad_args if grad is not None and grad.requires_grad])
+
+            # In order to keep map functional for backward graph,
+            # we clone outputs that are aliasing inputs
+            input_storage = {StorageWeakRef(arg._typed_storage()) for arg in flat_args if isinstance(arg, torch.Tensor)}
+
+            def maybe_clone(t):
+                if isinstance(t, torch.Tensor) and StorageWeakRef(t._typed_storage()) in input_storage:
+                    return t.clone()
+                return t
+            # flat list of grad_carry + grad_xs
+            return pytree.tree_map(maybe_clone, grads)
+
+        joint_graph = make_fx(joint_f)(*example_grad_carry_out, *example_grad_ys, *example_init, *example_xs)
+        return fw_graph, joint_graph
+
+
+def scan_wrapper(f, init, xs):
+    flat_init, init_spec = pytree.tree_flatten(init)
+    if not all(isinstance(t, torch.Tensor) for t in flat_init):
+        raise RuntimeError(f"Scanned init can only consist of tensors. Got init {flat_init}.")
+    flat_xs, xs_spec = pytree.tree_flatten(xs)
+    if not all(isinstance(t, torch.Tensor) for t in flat_xs):
+        raise RuntimeError(f"Scanned xs can only consist of tensors. Got xs {flat_xs}.")
+    
+    # TODO: Is it necessary to restrict the shape of scanned or mapped elements?
+    # for scan, the only requirement should be that f spits out a carry that
+    # always has the same shape
+    shapes = [xs.shape for xs in flat_xs]
+    leading_dim_size = shapes[0][0]
+    if leading_dim_size == 0:
+        raise RuntimeError(
+            "Leading dimensions of scanned xs cannot be 0.")
+
+    if any(cur_shape[0] != leading_dim_size for cur_shape in shapes):
+        raise RuntimeError(
+            f"Leading dimensions of scanned xs must be consistent. Got shapes {shapes}.")
+
+    carry_out_spec = None
+    out_spec = None
+    num_init_args = len(flat_init)
+
+    def flat_fn(*flat_args):
+        # carry and init should have the same spec
+        flat_init, flat_xs = flat_args[:num_init_args], flat_args[num_init_args:]
+        carry = pytree.tree_unflatten(flat_init, init_spec)
+        xs = pytree.tree_unflatten(flat_xs, xs_spec)
+        unflattened_carry_out, unflattened_out = f(carry, xs)
+        flat_carry_out, tmp_carry_out_spec = pytree.tree_flatten(unflattened_carry_out)
+        flat_out, tmp_out_spec = pytree.tree_flatten(unflattened_out)
+
+        nonlocal carry_out_spec
+        nonlocal out_spec
+        carry_out_spec = tmp_carry_out_spec
+        out_spec = tmp_out_spec
+        return *flat_carry_out, *flat_out
+    
+    flat_carry_out, flat_out = scan_impl(flat_fn, flat_init, flat_xs, return_all_carries=False, reverse=False)  # (ys, carry)
+    return pytree.tree_unflatten(flat_carry_out, carry_out_spec), pytree.tree_unflatten(flat_out, out_spec)
+
+class ScanAutogradOp(torch.autograd.Function):
```

Comment:
I get how this works in theory, but I need to re-derive what is going on every time I see it.

Is it possible to write a note with a small example somewhere so we can see what is going on?

Reply:
Here's how I derived the backwards of scan (and I'd love to see this expanded upon and put into the code as a note):
```
# scan(f, init, xs) -> carry, ys
# grad_carry, grad_ys -> grad_init, grad_xs
#
# Assume single iteration.
# f(init, xs) -> carry, ys
# In the forward pass:
# - f(init, xs) -> carry, ys
# In the backwards pass
# - joint_f(init, xs, grad_carry, grad_ys) -> carry, ys, grad_init, grad_xs
#
# Assuming two iterations
# f(init, xs[0]) -> carry0, ys0
# f(carry0, xs[1]) -> carry1, ys1
#
# We are given:
# grad_ys1, grad_ys0, grad_carry1
#
#
# joint_f(carry0, xs[1], grad_carry1, grad_ys1) -> grad_carry0, grad_xs1
# joint_f(init, xs[0], grad_carry0, grad_ys0) -> grad_init, grad_xs0
#
# So this is a scan(joint_f, init=grad_carry1, xs=(xs, grad_ys, carries), reverse=True)
```